Adding phe

wilko77 opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team](https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team) using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel](https://gitter.im/conda-forge/conda-forge.github.io)
or on our [Keybase chat](https://keybase.io/team/condaforge.chat)
if your recipe isn't reviewed promptly.
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/phe```) and found some lint.

Here's what I've got...


For **recipes/phe**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [14]

For **recipes/phe**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/phe```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipes/phe**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

hmaarrfk commented on recipes/phe/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+  commands:
+    - pheutil --help
+  requires:
+    - click
```
these seem like they should be runtime requirement. in the `run` section.

hmaarrfk commented on recipes/phe/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+
+about:
+  home: "https://github.com/data61/python-paillier"
+  license: "GNU General Public v3 (GPLv3)"
```
should be `GPL-3.0-only` or `GPL-3.0-or-later`

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/phe```) and found it was in an excellent condition.

wilko77 commented:
Thanks for the feedback @hmaarrfk. 
I changed the license string. 
With the requirements I'm not totally sure.
We wrote a pure python fallback into our library, thus there are no explicit requirements for using it.
There is a speedup if gmpy2 or pycrypto is available.  

The tests however, need those libraries to be able to execute all tests.

Is this approach right?
I left the 
```
test:
  requires:
    - stuff
```
section. I also added:
```
requirements:
  run_constrained:
    -gmpy2
```

hmaarrfk commented:
Since you have no constraint, then the run constrained might not be worth while.

The conda forge philosophy is more that your package should be free to include dependencies since they are now easy to install. That said, it also wants to give you freedom in how you decide to package hour software.

hmaarrfk commented:
This looks good to me. Somebody else will have to come by and merge thoigh

beckermr commented on recipes/phe/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+
+about:
+  home: "https://github.com/data61/python-paillier"
+  license: GPL-3.0-only
```
license is wrong since parts are apache

hmaarrfk replied:
which parts?

wilko77 replied:
Thanks for the feedback Matthew.
I was under the impression that it is okay to include Apache licensed software in GPL3 projects, as [stated by the Apache software foundation](http://www.apache.org/licenses/GPL-compatibility.html).

chrisburr replied:
It is, but to use the software you have to comply with both licences:
```suggestion
  license: GPL-3.0-only AND Apache-2.0
```

wilko77 replied:
Oh, I see. Thanks, Chris.

chrisburr commented on recipes/phe/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+    - python
+  run:
+    - python
+  run_constrained:
```
@systemshift  Does anything in `phe` work if `click`, `gmpy2` and `numpy` are not installed?

`run_constrained` is used for controlling the versions of packages so it won't actually do anything here if you don't have a version constraint.

systemshift replied:
@chrisburr  yeah I have read @hmaarrfk's comment regrading this. Unless @wilko77 has something to add, I made pr for him to accept https://github.com/wilko77/staged-recipes/pull/2.

chrisburr replied:
In that case the `requires` section under `tests` should also be removed. That section is only normally used for things like `pytest` that really are only needed for testing.

wilko77 replied:
I removed the run_constrained section, as it wasn't the right thing.

Our library has a pure python fall-back. However, if there are libraries like gmpy2 available, it will make use of some of the highly optimized functions for better performance.
In the `setup.py` we can specify this in the `extras_require` section. I didn't find an equivalent for the `meta.yaml` file. Thus, I just leave that out.

We do however test the functionality of using those optimized libraries, thus the tests require `click`, `gmpy2` and `numpy` to run.

I made changes to the meta.yaml file and hope that everything is in order now.
